swift/ql/lib/upgrades/f5a22f5168adfd7f308b5941c11852c925edd638/for_in_upgrade.ql

```diff
@@ -0,0 +1,261 @@
+}
+
+query predicate new_pattern_binding_decl_patterns(NewElement id, int index, NewElement pattern) {
+  pattern_binding_decl_patterns(id, index, pattern)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TIteratorVar(foreach)) = id and
+    Fresh::map(TIteratorVarPattern(foreach)) = pattern and
+    index = 0
+  )
+}
+
+query predicate new_named_patterns(NewElement id, string name) {
+  named_patterns(id, name)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TIteratorVarPattern(foreach)) = id and
+    name = foreach.getPattern().getGeneratorString()
+  )
+}
+
+query predicate new_pattern_binding_decl_inits(NewElement id, int index, NewElement init) {
+  pattern_binding_decl_inits(id, index, init)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TIteratorVar(foreach)) = id and
+    foreach.getSequence() = init and
+    index = 0
+  )
+}
+
+query predicate for_each_stmt_next_calls(ForEachStmt id, NewElement nextCall) {
+  Fresh::map(TNextCall(id)) = nextCall
+}
+
+query predicate new_dot_syntax_call_exprs(NewElement id) {
+  dot_syntax_call_exprs(id)
+  or
+  Fresh::map(TNextCallMethodLookup(_)) = id
+}
+
+query predicate new_self_apply_exprs(NewElement id, NewElement base) {
+  self_apply_exprs(id, base)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallMethodLookup(foreach)) = id and
+    Fresh::map(TNextCallInOutConversion(foreach)) = base
+  )
+}
+
+query predicate new_in_out_exprs(NewElement inOutExpr, NewElement subExpr) {
+  in_out_exprs(inOutExpr, subExpr)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallInOutConversion(foreach)) = inOutExpr and
+    Fresh::map(TNextCallVarRef(foreach)) = subExpr
+  )
+}
+
+query predicate new_apply_exprs(NewElement id, NewElement func) {
+  apply_exprs(id, func)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCall(foreach)) = id and
+    Fresh::map(TNextCallMethodLookup(foreach)) = func
+  )
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallMethodLookup(foreach)) = id and
+    Fresh::map(TNextCallFuncRef(foreach)) = func
+  )
+}
+
+Element getParent(Element type) {
+  any_generic_type_parents(type, result)
+  or
+  // there may be an extension that implements IteratorProtocol
+  exists(@extension_decl extDecl, @nominal_type_decl typeDecl, @protocol_decl protocolDecl |
+    any_generic_types(type, typeDecl) and
+    extension_decls(extDecl, typeDecl) and
+    extension_decl_protocols(extDecl, _, protocolDecl) and
+    any_generic_types(result, protocolDecl)
+  )
+}
+
+Element getNextMethod(ForEachStmt foreach) {
+  exists(@element sequence, @element exprType, @element parentType, @element typeDecl |
+    sequence = foreach.getSequence() and
+    expr_types(sequence, exprType) and
+    parentType = getParent*(exprType) and
+    any_generic_types(parentType, typeDecl) and
+    decl_members(typeDecl, _, result) and
+    callable_names(result, "next()")
+  )
+}
+
+// TODO: do we need a new_apply_expr_arguments
```

Comment:
What's the status of this TODO?

Reply:
We don't need it - the `this` parameter is part of `new_self_apply_exprs`.